File: util/convertcsv2snana.pl
#!/usr/bin/perl
#
# Jan 2014 R.Kessler
#
# Translate csv formatted file into a file with
# the snana/fitres format:
#  VARNAMES: ROW <VARNAME1> <VARNAME2> <VARNAME3> ...
#  ROW: 1  <value1> <value2> <value3> ...
#  ROW  2  <value1> <value2> <value3> ...
#  etc ...
#
# If the first variable name is NOT CID or ROW or STARID, 
# then a ROW column is inserted.  Either comma or blank-space 
# delimeters are allowed.
#
#
# Usage
#  convertcsv2snana.pl <inFile>
#
#  convertcsv2snana.pl <inFile> CAPS
#       [all varnames -> uppercase]
#
#  convertcsv2snana.pl <inFile> CAPS   --VARNAMES <list of varnames>
#     [replace varnames list]
#
#
# May 12 2016: fix to allow '#' in first line.
# Jun 27 2019: remove NVAR key that is obsolete.
# ----------------------

use strict ;

my ($INFILE, $OPT_CAPS, @VARLIST_USER) ;
my ($OUTFILE, $TMPFILE, $SEDCMD);
my ($NVAR, $VARNAMES_ORIG, @VARLIST_ORIG, $VARNAMES_OUT);

sub parse_args ;
sub get_VARNAMES  ;
sub make_outFile ;

# ============== START MAIN ===============

&parse_args();


# remove commas and comment signs.
# ubstitute up to two ,, with ,-9999' in case of blanks; 
# this does not protect blank in last column.
$SEDCMD = "sed " .
    "-e 's/,,/,-9999,/g' " .
    "-e 's/,,/,-9999,/g' " . 
    "-e 's/,/ /g' " .
    "-e 's/#//g'" ; 
qx($SEDCMD $INFILE > $TMPFILE);

#  list of varnames
&get_VARNAMES();

&make_outFile();

print "\n Done. See $OUTFILE \n";

# =========== END MAIN =============


sub parse_args {
    
    my ($i, $NARG, $USERFLAG );

    $OPT_CAPS = 0 ;
    @VARLIST_USER = ()  ;
    $USERFLAG = 0 ;

    $NARG = scalar(@ARGV);

    if ( $NARG < 1 ) {
	die "\n ERROR: must specify input file \n";
    }

    
    $INFILE  = $ARGV[0];
    $OUTFILE = "${INFILE}_converted" ;
    $TMPFILE = "TMP_${INFILE}" ;

    for($i=1; $i < $NARG; $i++ ) {
	if ( $ARGV[$i] eq "CAPS"       ) { $OPT_CAPS = 1; }

	if ( $USERFLAG ) { @VARLIST_USER = (@VARLIST_USER, $ARGV[$i]); }
	if ( $ARGV[$i] eq "--VARNAMES" ) { $USERFLAG = 1; }
	if ( $ARGV[$i] eq "-VARNAMES"  ) { $USERFLAG = 1; }
    }

    print "\n";
    print " Translate $INFILE  -> $OUTFILE \n";


}  # end of parse_args


# =====================================
sub get_VARNAMES {

    open PTR_INFILE , "< $INFILE" ;
    $VARNAMES_ORIG = <PTR_INFILE> ;
    close PTR_INFILE ;

    if ( $OPT_CAPS ) { $VARNAMES_ORIG = uc($VARNAMES_ORIG); }

    # replace commas in VARNAMES with blank spaces so that
    # either comma or space delimeter is allowed
    my ($VARNAMES_noComma);
    $VARNAMES_noComma = $VARNAMES_ORIG ;  
    $VARNAMES_noComma =~ s/,/ /g ;
    $VARNAMES_noComma =~ s/#//g ;  # remove comment sign (May 2016)

    @VARLIST_ORIG =  split(" ", $VARNAMES_noComma);

    $NVAR = scalar(@VARLIST_ORIG) ;

    print " VARNAMES_ORIG($NVAR) = @VARLIST_ORIG \n";

    $VARNAMES_OUT = "@VARLIST_ORIG" ;

    # check for user varnames
    my $NVAR_USER = scalar(@VARLIST_USER);
    if ( $NVAR_USER > 0 ) {
	
	if ( $NVAR_USER != $NVAR ) 
	{ die "\n ERROR: NVAR=$NVAR but NVAR_USER=$NVAR_USER \n"; }

	$VARNAMES_OUT = "@VARLIST_USER" ;	
    }

} # end of get_VARNAMES 


# ========================
sub make_outFile {

    # transfer contents to OUTFILE, and add ROW variable
    # in first column so that 'combine_fitres.exe' works.

    my (@CONTENTS, $NROW, $NLINE, $LINE );

    # read everything from input file
    open PTR_TMPFILE , "< $TMPFILE" ;
    @CONTENTS = <PTR_TMPFILE> ;
    close PTR_TMPFILE ;
    qx(rm $TMPFILE);
    
    my ($FIRSTVAR, $VARNAME_ROW, $ADDROW, $cNROW);
    $ADDROW      = 1 ;
    $FIRSTVAR    = $VARLIST_ORIG[0];

    if ( $FIRSTVAR eq "CID"    ) { $ADDROW = 0 ; }
    if ( $FIRSTVAR eq "cid"    ) { $ADDROW = 0 ; }
    if ( $FIRSTVAR eq "ROW"    ) { $ADDROW = 0 ; }
    if ( $FIRSTVAR eq "GALID"  ) { $ADDROW = 0 ; }
    if ( $FIRSTVAR eq "STARID" ) { $ADDROW = 0 ; }

    if ( $ADDROW ) 
    { $VARNAME_ROW = "ROW" ; $NVAR = $NVAR + 1 ; }
    else
    { $VARNAME_ROW = "" ; }
	

    # open  OUTFILE
    open PTR_OUTFILE , "> $OUTFILE" ;
    
# xxx mark delete Jun 2019  print PTR_OUTFILE "NVAR: $NVAR \n";
    print PTR_OUTFILE "VARNAMES: $VARNAME_ROW $VARNAMES_OUT \n";

    $NROW = $NLINE = 0 ;
    foreach $LINE (@CONTENTS) {
	$NLINE++ ;
	if ( $NLINE == 1 ) { next ; } # skip variable definitions
	
	$LINE =~ s/\s+$// ; # remove trailing blanks
	$NROW++ ;	
	$cNROW = sprintf("%3d", $NROW); 

	if ( $ADDROW ) 
	{  print PTR_OUTFILE "ROW: $cNROW  $LINE \n"; }
	else
	{  print PTR_OUTFILE "ROW: $LINE \n"; }
    }

    print " Finished writing $NROW rows to: $OUTFILE \n";
    close PTR_OUTFILE ;

} # end of make_outFile

File: util/convertcsv2snana.py
#!/usr/bin/env python
#
# Translate csv format into snana-key format.
# If fisrt column is CID,GAL,STARTID,ROW ...,  then translate
# without modification; otherwise add ROW column for 
# SNANA-KEY format
#
# Apr 27 2021: replace '' with VOID (works only with first '')
# Mar 05 2024: 
#  + write "GAL:" row key if GALID is in header
#  + fix to work with either comma-sep or space-sep var list in header.
#
# Dec 01 2025: 
#   + replace blanks and dash (-) with -9
#   + if valid row key (e.g. SNID) is not in column 0, them automatically move it
#     to row 0
#
import os, sys, argparse, csv

VALID_ROWID_LIST = [ "CID", "SNID", "snid", "ROW", "GALID", "STARID" ]
ROWID_DEFAULT    = "ROW" 

ROWKEY = "ROW:"  # default row key

BLANK_LIST = [ '', '-' ]  # replace this nulls with VALUE_BLANK
VALUE_BLANK = '-9'   # replace blanks with this value

# =====================================
def get_args():
    parser = argparse.ArgumentParser()
    
    msg = "name of input csv file"
    parser.add_argument("-i", "--csv_file", help=msg, type=str, default=None)

    msg = "optional name of output text file"
    parser.add_argument("-o", "--out_file", help=msg, type=str, default=None)

    msg = "write output VARNAMES with all CAPS"
    parser.add_argument("-C", "--CAPS", help=msg, action="store_true")

    args = parser.parse_args()

    if len(sys.argv) == 1:  parser.print_help(); sys.exit()

    if args.csv_file is None :
        sys.exit("ERROR: must specifiy input csv file")

    return args

# ====================================
def get_out_file_name(args):
    if args.out_file is not None :
        out_file = args.out_file
    else:
        infile   = (args.csv_file).split(".")[0]
        out_file = infile + ".text"
        
    print(f" Write SNANA-key format to {out_file}")
    return out_file

# =================================
def read_csv_file(csv_file):

    print(f" Read csv contents from: {csv_file}")

    nrow = 0; row_list = []
    with open(csv_file, 'r') as f:  
        contents = csv.reader(f)

        for row in contents :
            if len(row) == 0 : continue
            if row[0][0] == '#' : continue
            if nrow == 0 :
                varname_list = row
            else:
                row_list.append(row)
            nrow += 1

            if '' in row:
                j = row.index('')
                row[j] = 'VOID'

    #print(f" xxx varname_list = {varname_list}")
    #print(f" xxx row_list = {row_list}")

    # - - - - - -                                                               
    # if varname_list has only one item,                                        
    #     e.g., [ 'GALID ZPHOT ZPHOTERR' ], then split it into 
    # list of variables. This will enable reading both comma-sep 
    # and space-sep list of variables in header.                   
    if len(varname_list) == 1 :
        varname_list = varname_list[0].split()

    return varname_list, row_list
    # end read_csv_file

def get_colnum_rowid(varname_list):

    colnum = -9
    for valid_rowid in VALID_ROWID_LIST:
        if valid_rowid in varname_list:
            colnum = varname_list.index(valid_rowid)

    return colnum

# ==============================
def  write_out_file(out_file,varname_list,row_list):

    # check if first varname is a valid IDentifier

    add_rownum = False 
    # xxx varname0 = varname_list[0]

    colnum_rowid = get_colnum_rowid(varname_list)

    if colnum_rowid >= 0:
        
        varname_rowid = varname_list[colnum_rowid]
        if varname_rowid == "GALID": 
            global ROWKEY ; ROWKEY = "GAL:"

        if colnum_rowid == 0:
            header = varname_list
        else:
            print(f" Move {varname_rowid} column from colnum={colnum_rowid} to colnum=0")
            rowid =  varname_list.pop(colnum_rowid)
            del      varname_list[colnum_rowid] 
            header = [ rowid ] + varname_list
            #sys.exit(f"\n xxx header = \n{header}")
    else:
        header = [ ROWID_DEFAULT ] + varname_list
        add_rownum = True


    f = open(out_file,"wt")

    # write command in comment fields
    command = " ".join(sys.argv)
    f.write(f"# Created with command\n")
    f.write(f"#   {command} \n#\n")

    # write list of VARNAMES
    line = "VARNAMES: " + " ".join(header)
    f.write(f"{line}\n")

    # write each row
    rownum = 0 
    for row_orig in row_list:

        row = row_orig
        rownum += 1
        line  = f"{ROWKEY} "
        if add_rownum : 
            line += (f"{rownum:3d}  ")

        if colnum_rowid > 0:  # move rowid value to 1st column
            rowid =  row.pop(colnum_rowid)
            del      row[colnum_rowid] 
            row    = [ rowid ] + row

        for blank in BLANK_LIST:
            if blank in row: row = [ VALUE_BLANK  if x == blank else x for x in row]

        line += " ".join(row)
        f.write(f"{line}\n")

    f.close()


# =========================
# ======= MAIN ============
# =========================

if __name__ == "__main__":

    # read command-line arguments
    args  = get_args()

    # read contents of csv file
    varname_list, row_list = read_csv_file(args.csv_file)

    out_file = get_out_file_name(args)

    # check all caps option
    if args.CAPS :
        VARNAME_LIST = [v.upper() for v in varname_list ]
    else:
        VARNAME_LIST = varname_list

    write_out_file(out_file, VARNAME_LIST, row_list)
    print(" Done.\n")

# END
